oauth-keycloak-authorizer/src/main/java/io/strimzi/kafka/oauth/server/authorizer/GrantsHandler.java

```diff
@@ -0,0 +1,520 @@
+            }
+
+            Set<Map.Entry<String, Info>> entries = workmap.entrySet();
+            List<Future> scheduled = new ArrayList<>(entries.size());
+            for (Map.Entry<String, Info> ent : entries) {
+                String userId = ent.getKey();
+                Info grantsInfo = ent.getValue();
+                if (grantsInfo.getLastUsed() < System.currentTimeMillis() - grantsMaxIdleMillis) {
+                    log.debug("Skipping refreshing grants for user '{}' due to max idle time.", userId);
+                    removeUserFromCacheIfExpiredOrIdle(userId);
+                }
+                scheduled.add(new Future(userId, grantsInfo, refreshWorker.submit(() -> {
+
+                    if (log.isTraceEnabled()) {
+                        log.trace("Fetch grants for user: " + userId + ", token: " + mask(grantsInfo.getAccessToken()));
+                    }
+
+                    JsonNode newGrants;
+                    try {
+                        newGrants = fetchGrantsWithRetry(grantsInfo.getAccessToken());
+                    } catch (HttpException e) {
+                        if (403 == e.getStatus()) {
+                            // 403 happens when no policy matches the token - thus there are no grants, no permission granted
+                            newGrants = JSONUtil.newObjectNode();
+                        } else {
+                            throw e;
+                        }
+                    }
+                    JsonNode oldGrants = grantsInfo.getGrants();
+                    if (!semanticGrantsEquals(newGrants, oldGrants)) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Grants have changed for user: {}; before: {}; after: {}", userId, oldGrants, newGrants);
+                        }
+                        grantsInfo.setGrants(newGrants);
+                    }
+
+                    // Only added here to allow compiler to resolve the lambda as a Callable<JsonNode>
+                    return newGrants;
+                })));
+            }
+
+            for (GrantsHandler.Future f : scheduled) {
+                try {
+                    f.get();
+                } catch (ExecutionException e) {
+                    final Throwable cause = e.getCause();
+                    if (cause instanceof HttpException) {
+                        log.debug("[IGNORED] Failed to fetch grants for user: {}", cause.getMessage());
+                        if (401 == ((HttpException) cause).getStatus()) {
+                            grantsCache.remove(f.getUserId());
+                            log.debug("Removed user from grants cache: {}", f.getUserId());
+                            Services.getInstance().getSessions().removeAllWithMatchingAccessToken(f.getGrantsInfo().accessToken);
+                            continue;
+                        }
+                    }
+                    if (log.isWarnEnabled()) {
```

Comment:
Is the call to `e.getMessage()` so expensive that this guard is worth the noise?